### DISCUSSION TITLE
close before stop or stop before close in embedded etcd?

### DISCUSSION THREAD

**Discussion by bluebrown** [etcd-io/etcd#15724] [Q&A] [answered]
Hi, I am playing around with https://pkg.go.dev/go.etcd.io/etcd/server/v3/embed. I am confused by the [etcd.Close](https://pkg.go.dev/go.etcd.io/etcd/server/v3/embed#Etcd.Close) method and [etcdserver.Stop](https://pkg.go.dev/go.etcd.io/etcd/server/v3@v3.5.8/etcdserver#EtcdServer.Stop) method. 

When shutting down the program, in which order are they supposed to be called?

i.e.  like below, or the other way around?

```go
e, _ := embed.StartEtcd(cfg)
...
e.Server.Stop()
e.Close()
return <-e.Err()
```

From the example here it seems like stop comes first, since the close call is deferred. https://github.com/etcd-io/etcd/blob/main/server/embed/doc.go#L27

But from my testing, e.Err may block until close is actually called. So deferring it may lead to the program hanging. For example see this https://go.dev/play/p/Trmcy9TV4ag.

**Comment by tjungblu** [comment, marked as answer]
If you look at the code, the `Close` method will also stop:
https://github.com/etcd-io/etcd/blob/server/v3.5.8/server/embed/etcd.go#L423-L426

So it's enough to just Close.

**Comment by ahrtr** [comment]
I agree with @tjungblu . User application just needs to call `Close()`, and it's recommended to follow the same way as etcdserver talks to the embed package. FYI. https://github.com/etcd-io/etcd/blob/b27dec8b9487d0d9358ca4dd366563d1aab04a1e/server/etcdmain/etcd.go#L185-L199

Also please feel free to raise a PR to fix the incorrect [doc](https://github.com/etcd-io/etcd/blob/main/server/embed/doc.go#L27-L43).
